velox/exec/PartitionedOutputBufferManager.h

```diff
@@ -21,12 +21,29 @@
 
 namespace facebook::velox::exec {
 
-// nullptr in pages indicates that there is no more data.
-// sequence is the same as specified in BufferManager::getData call. The caller
-// is expected to advance sequence by the number of entries in groups and call
-// BufferManager::acknowledge.
-using DataAvailableCallback = std::function<
-    void(std::vector<std::unique_ptr<folly::IOBuf>> pages, int64_t sequence)>;
+enum GetDataStatus { SUCCESS, ERR_BUFFER_NOT_FOUND };
+
+class DataAvailableCallbackParam {
```

Comment:
Use struct.

This change is not backward compatible and will break Prestissimo. Check with @aditi-pandit about strategies for making backwards-compatible changes.

We can keep the existing callback API, but modify the signature of getData API to return bool instead of void. The method would return true if it successfully queued the 'notify' callback to fire when data becomes available. It will return false if 'notify' callback cannot be installed because there are no output buffers for the specified task (either because these never existed or have been already deleted upon task failure).

Reply:
@karteekmurthys Breaking changes are no longer allowed (CC: @spershin). Hence, we would need to find a way to make a change without breaking Prestissimo.

> notify() and bool (as return value), the getData communicates with Caller in 2 ways. I find this a bit confusing.

Yes, this makes the API a bit more complex, but I think that's Ok. We just need to document it clearly. The benefit of returning 'false' instead of calling notify with some special status is less overhead for handling a corner case of when Task failed very quickly after Task::start call and before the POBM::getData call. It also makes the logic of 'notify' callback simplier.

Reply:
Thanks @mbasmanova. Karteek and I discussed this morning about backward compatibility and this cleaner API.